### PR TITLE
Collect miniapps/common/*.o in a library file [lib-extras-dev]

### PR DESCRIPTION
**Pull request by mlstowell**
The purpose of this PR is to package the object files in miniapps/common within a library file that user applications can link with if needed.  These object files contain convenience classes and functions that don't need to be part of `libmfem` put may be useful for developers who use existing miniapps as starting points for building their own applications.

This PR adds a common header file called `miniapps/common/mfem-common.hpp` and build targets to create `miniapps/common/libmfem-common.a`.

Changes will also be needed in the CMake files but I wanted to get a little feedback on the general idea first.

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/864 | @mlstowell | @tzanio  | @tzanio + @rcarson3 | 12/15/19 | 01/08/20 | ⌛due 01/15/20 |

### CODE REVIEW

**Comment by tzanio** [comment]
I'm not really convinced we need another library... at least not at the top level.

@v-dobrev what do you think?

**Comment by rcarson3** [comment]
So currently when ```cmake``` is used and  ```make install``` is called, the miniapp directory stays within the build directory. Would it make sense @tzanio, @v-dobrev, and @mlstowell  with this PR to have a miniapp directory with this library and the necessary headers included in the install directory?

**Comment by jandrej** [comment]
This has been discussed offline and will be introduced in a future PR.

**Comment by mlstowell** [comment]
Hi, @rcarson3 ,
The intention is to make the `libmfem-extras` library available to outside developers in a manner similar to how `libmfem` is provided.  The new library (and its headers) doesn't need to be in the same location as `libmfem` but it should be installed somewhere convenient.  I know @tzanio has stronger feelings about where it belongs than I do so I'll defer to him.
Thanks,
Mark

**Comment by v-dobrev** [comment]
The new `libmfem-extras.a` does not have a lot of code, so it seems simpler (for users) to merge it into `libmfem.a`. Naturally, the header `mfem-extras.hpp` will also be included by `mfem.hpp`.

Also, currently the additional code is in the namespace `mfem::miniapps` -- it seems more appropriate to use `mfem::extras` or, alternatively, rename the header to `mfem-miniapps.hpp`.

**Comment by mlstowell** [comment]
@v-dobrev ,
There is certainly an argument to be made for simply putting these things in `libmfem.a` (@tzanio made the same point).  Part of the reason for keeping them separate is that these things don't generally add to the functionality of MFEM, as a finite element library, they mainly provide convenient wrappers which can be handy for application developers.  We may add various convenience classes to `libmfem-extras.a` over time and we probably don't want these to cause `libmfem.a` to bloat.  Another way to think about it is that `libmfem.a` provides hard-core finite element code in full detail whereas `libmfem-extras.a` offers a handful of shortcuts for commonly constructed objects or similar pieces of code (like `VisualizeField`).

Using the namespace `mfem::extras` does make more sense, good point!

File: .gitignore [hunk 1]
```diff
@@ -9,6 +9,7 @@
 # Object and library files
 *.o
 /libmfem.*
+/libmfem-extras.*
```
Comment (tzanio) [inline .gitignore]: This should have a `miniapps` prefix

File: makefile [hunk 2]
```diff
@@ -515,7 +538,8 @@ $(ALL_CLEAN_SUBDIRS):
 	$(MAKE) -C $(BLD)$(@D) $(@F)
 
 clean: $(addsuffix /clean,$(EM_DIRS) $(TEST_DIRS))
-	rm -f $(addprefix $(BLD),*/*.o */*~ *~ libmfem.* deps.mk)
+	rm -f $(addprefix $(BLD),*/*.o */*~ *~ libmfem.* \
+	miniapps/libmfem-extras.* deps.mk)
```
Comment (tzanio) [inline makefile]: the `install` target has to be modified

**Comment by tzanio** [comment]
I made a few comments on the code, but I am still conflicted about the role of this library.

If the aim is to help the folks that will be looking at the miniapps, then I think both the header and the library should be there. It will also be useful to use more descriptive names, e.g.

- `miniapps/common` ➡️ `miniapps/utilities`
- `miniapps/common/*_extras.*` ➡️ `miniapps/utilities/utilities.`
- `mfem-extras.hpp` ➡️ `miniapps/mfem-utilities.hpp`
- `miniapps/libmfem-extras.a` ➡️ `miniapps/libmfem-utilities.a`

(instead of `utilities`, other naming options could be `utils`,`helpers`, `app-kit`, or just `kit`)

**Comment by mlstowell** [comment]
Considering `utilities`, or something derived therefrom, versus `extras` I came across something that made me chuckle.

According to Merriam-Webster:
utility (advective): ... serving primarily for utility rather than beauty.
extra (noun): ... an attractive addition or accessory.

Now which do you think I prefer?  Seriously, I do prefer "extras" because I think it makes it plain that these are not strictly necessary.  People can make full use of MFEM without ever looking at the "extras" and they don't even need to link with them if they don't want to.

**Comment by tzanio** [comment]
This could be easier to discuss in person, but I think my confusion comes from the association of the "extras" with the "miniapps".

If we want people to think of these as "extra tools/utilities not included in the main mfem library", then the code should not be in the `miniapps/common` directory and we should consider instead adding a new top-level directory, `extras` for it.

If we want to keep the code in `miniapps/common` as is, then the mixing of "extras" + "miniapps" is just confusing for me. Picking a name to indicate that these are application-building tools will be more clear.

**Comment by mlstowell** [comment]
@tzanio , that does make a lot of sense.

**Comment by mlstowell** [comment]
@tzanio and I just spoke at length about the naming and placement of this new library and I think I better understand his position.  One of the main goals is to avoid confusing our users.  With that in mind we'll keep this PR's changes confined to the `miniapps` subdirectories.  This means moving `mfem-extras.hpp` into either `miniapps` or `miniapps/common`.  The new library will be created in the same location i.e. either `miniapps` or `miniapps/common`.  Which location should be chosen?

The other issue is naming. I think we agree that the names should be of the form `mfem-something.hpp` and `libmfem-something.a`.  What should we choose in place of `mfem-something`?
`mfem-app-kit`
`mfem-common`
`mfem-extras`
`mfem-miniapps`
`mfem-utilities`
... other suggestions?

**Comment by tzanio** [comment]
I'd like to hear from @v-dobrev but here are my personal preferences...

For the location (in order of preference)
1. `miniapps/common`
2. `miniapps`
3. `miniapps/common` renamed as `miniapps/extras` or `miniapps/utilities`

For the name (in order of preference)
1. `mfem-common` in `miniapps/common`
2. `mfem-miniapps` in `miniapps`
3. `mfem-extras` in `miniapps` or in `miniapps/common` renamed as `miniapps/extras`
4. `mfem-utilities` in `miniapps` or in `miniapps/common` renamed as `miniapps/utilities`
5. `mfem-app-kit` in `miniapps`

**Comment by jandrej** [comment]
Sorry for chiming in here, but do we want to consider future developments here? I'm thinking of our discussion regarding the navier stokes library and its directory in the repository as well as in the installation directory.

**Comment by tzanio** [comment]
This is blocking `toys-dev`, which I think is otherwise ready for review... can we maybe meet to resolve it?

**Comment by mlstowell** [comment]
Any afternoon next week (I'm on vacation this week) would be fine with me.

**Comment by tzanio** [comment]
After discussion in person, we agreed with @mlstowell on the library name `miniapps/common/libmfem-common.so/a` and the header name `miniapps/common/mfem-common.h`

**Comment by tzanio** [comment]
Merged in `next` for testing ...

**Comment by v-dobrev** [comment]
Looking at the current state, I do not like that there is a separate library that is being installed. This makes it more difficult for users to link and makes it harder for developers to maintain.

If we **really** need to have separate libraries then we need to have a technical meeting to discus what's the best way to do this. For example, I think such additional libraries should be optional. Also, I think it is much more natural for users to have the extra library files and headers next to the main mfem library and headers, respectively.

**Comment by tzanio** [comment]
@mlstowell -- I spoke with @v-dobrev and one possible compromise is to keep things as they are now, except revert the makefile not to build or install the `mfem-common` library (this will also resolve the regression test errors).

Are you OK with this change?

If not, maybe the three of us can meet tomorrow afternoon to discuss this?

**Comment by mlstowell** [comment]
Hi, @tzanio,

I don't think I understand the proposal.  If the makefile doesn't build the library then this PR would only be adding a header file.  I'm also curious about the problem with the regression tests.  We should probably meet.
I'm sorry this has become such a hassle.  I really thought this was a trivial suggestion and I cannot believe we have spent so much time debating it.

Best wishes,
Mark

**Comment by v-dobrev** [comment]
Mark, I updated the build system as we discussed. Let me know if something does not look right.

**Comment by mlstowell** [comment]
It looks good to me.  Thanks, @v-dobrev !

**Comment by tzanio** [comment]
The following error was detected by the regression testing:

```
/usr/bin/ld: mesh_extras.o: relocation R_X86_64_32S against symbol `_ZTVSt15basic_streambufIcSt11char_traitsIcEE@@GLIBCXX_3.4' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: fem_extras.o: relocation R_X86_64_32S against undefined symbol `_ZTVN4mfem18FiniteElementSpaceE' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: pfem_extras.o: relocation R_X86_64_32S against symbol `_ZTVN4mfem6common13H1_ParFESpaceE' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Nonrepresentable section on output
make[1]: *** [libmfem-common.so.4.0.1] Error 1
make: *** [miniapps/common] Error 2
make: *** Waiting for unfinished jobs....
```

See the [mystash log](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/78d60374e4f0b7cefc20e0910c5087b57a464ca6#toss_3_x86_64_ib/2020-01-11-llnl-gh-next/build-test-tpls.out) for details.

**Comment by v-dobrev** [comment]
The shared library issue on linux should be fixed now.

**Comment by mlstowell** [comment]
Thanks, @v-dobrev !  I really appreciate your help with this one.

**Comment by tzanio** [comment]
With this merge, `toys-dev` can now be updated and I think will be ready for review.

**Pull request merged**